File: android/app/src/main/java/com/mochi/keyboard/features/home/HomeViewModel.kt
package com.mochi.keyboard.features.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mochi.keyboard.data.ThemeCache
import com.mochi.keyboard.data.ThemeRepository
import com.mochi.keyboard.model.KeyboardTheme
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed interface HomeUiState {
    data object Loading : HomeUiState
    data class Data(val recentlyApplied: List<KeyboardTheme>, val popular: List<KeyboardTheme>) : HomeUiState
    data object Empty : HomeUiState
    data class Error(val message: String) : HomeUiState
}

class HomeViewModel(private val themeRepository: ThemeRepository) : ViewModel() {

    private val _uiState = MutableStateFlow<HomeUiState>(HomeUiState.Loading)
    val uiState: StateFlow<HomeUiState> = _uiState.asStateFlow()

    init {
        load()
    }

    fun load() {
        viewModelScope.launch {
            _uiState.value = HomeUiState.Loading
            runCatching {
                val recentlyApplied = themeRepository.getPublishedThemes(limit = 3)
                val popular = themeRepository.getTopRanked(limit = 10)
                ThemeCache.put(recentlyApplied + popular)
                recentlyApplied to popular
            }.onSuccess { (recentlyApplied, popular) ->
                _uiState.value = if (recentlyApplied.isEmpty() && popular.isEmpty()) {
                    HomeUiState.Empty
                } else {
                    HomeUiState.Data(recentlyApplied, popular)
                }
            }.onFailure { e ->
                _uiState.value = HomeUiState.Error(e.message ?: "Couldn't load themes.")
            }
        }
    }
}

File: ios/MochiApp/Features/Home/HomeViewModel.swift
import Foundation

enum HomeUiState {
    case loading
    case data(recentlyApplied: [KeyboardTheme], popular: [KeyboardTheme])
    case empty
    case error(String)
}

/// Swift mirror of android/.../features/home/HomeViewModel.kt — same Loading/Data/Empty/Error shape,
/// same MockData-fallback convention every other screen in this app follows (Loading/Error render
/// the existing mock content rather than a spinner/error view designed nowhere in the Figma export;
/// only a genuinely empty catalog shows real empty state).
@MainActor
final class HomeViewModel: ObservableObject {
    @Published private(set) var uiState: HomeUiState = .loading

    private let themeRepository: ThemeRepository?

    /// `nil` container (no GoogleService-Info.plist yet) means there's no real backend to query at
    /// all — same as every other screen, this degrades to permanently showing MockData rather than
    /// attempting a Firestore call that would crash without configuration.
    init(container: AppContainer?) {
        themeRepository = container?.themeRepository
        load()
    }

    func load() {
        guard let themeRepository else { return }
        uiState = .loading
        Task {
            do {
                async let recentlyAppliedTask = themeRepository.getPublishedThemes(limit: 3)
                async let popularTask = themeRepository.getTopRanked(limit: 10)
                let (recentlyApplied, popular) = try await (recentlyAppliedTask, popularTask)
                uiState = (recentlyApplied.isEmpty && popular.isEmpty)
                    ? .empty
                    : .data(recentlyApplied: recentlyApplied, popular: popular)
            } catch {
                uiState = .error(error.localizedDescription)
            }
        }
    }
}
